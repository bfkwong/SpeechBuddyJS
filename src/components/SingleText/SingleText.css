.TopLevelDisplay {
  padding: 15px 25px 10px 25px;
  text-align: center;
}

.main-text-area {
  background-color: #cbe7f6 !important;
  border: none !important;
  border-radius: 10px !important;
}

.analyze-button {
  margin: 20px 0px 10px 0px;
  width: 100%;
  background-color: #3587a4 !important;
  border-color: #3587a4 !important;
}
