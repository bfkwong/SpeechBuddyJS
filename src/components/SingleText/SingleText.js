import React, { useState } from "react";
import natural from "natural";
import * as toxicity from "@tensorflow-models/toxicity";
import SpeechRecognition, {
  useSpeechRecognition
} from "react-speech-recognition";
import {
  Container,
  Row,
  Col,
  InputGroup,
  FormControl,
  Button,
  ButtonGroup,
  ButtonToolbar,
  Modal,
  Spinner
} from "react-bootstrap";
import "./SingleText.css";

import StatCard from "../StatCard/StatCard";
import SingleTextAnalysis from "../SingleTextAnalysis/SingleTextAnalysis";

import RecordingGif from "../../Images/recording.gif";

function SingleText() {
  const [textArea, setTextArea] = useState("");
  const [wordCount, setWordCount] = useState(0);
  const [sentenceCount, setSentenceCount] = useState(1);
  const [nounsPerc, setNounsPerc] = useState(0);
  const [adjPerc, setAdjPerc] = useState(0);
  const [textAnalysis, setTextAnalysis] = useState({});
  const [recording, setRecording] = useState(false);
  const [analysisState, setAnalysisState] = useState("DONE");
  const { transcript, resetTranscript } = useSpeechRecognition();

  const runAnalysis = async () => {
    setAnalysisState("Loading Models ...");
    let wordTok = new natural.WordTokenizer();
    let toxModel = await toxicity.load();
    let tokenizedWords = wordTok.tokenize(textArea);

    let wrdCnt = {};
    tokenizedWords.forEach((word) => {
      if (wrdCnt[word] === undefined) {
        wrdCnt[word] = 0;
      }
      wrdCnt[word] += 1;
    });
    let wordCountPair = Object.keys(wrdCnt).map((key) => [key, wrdCnt[key]]);
    wordCountPair.sort(function (a, b) {
      // Sort by the 2nd value in each array
      if (a[1] === b[1]) return 0;
      return a[1] < b[1] ? 1 : -1;
    });
    let topTenWords =
      wordCountPair.length > 20 ? wordCountPair.slice(0, 20) : wordCountPair;

    setAnalysisState("Analyzing Sentiment ...");
    let analyzer = new natural.SentimentAnalyzer(
      "English",
      natural.PorterStemmer,
      "afinn"
    );

    setAnalysisState("Analyzing for Toxicity ...");
    if (toxModel) {
      let predictions = await toxModel.classify([textArea]);
      if (!predictions) {
        alert("prediction failed");
        return;
      }
      setTextAnalysis({
        sentiment: analyzer.getSentiment(tokenizedWords),
        toxicity: predictions,
        topTenWords: topTenWords
      });
      setAnalysisState("DONE");
    }
  };

  return (
    <div className="TopLevelDisplay">
      <h1>Single Text Analysis</h1>
      <Container>
        <Row className="justify-content-md-center">
          <Col md={3} xs={6}>
            <StatCard value={wordCount} description="Words"></StatCard>
          </Col>
          <Col md={3} xs={6}>
            <StatCard value={sentenceCount} description="Sentences"></StatCard>
          </Col>
          <Col md={3} xs={6}>
            <StatCard
              value={`${Math.round(nounsPerc * 100)}%`}
              description="Are Nouns"></StatCard>
          </Col>
          <Col md={3} xs={6}>
            <StatCard
              value={`${Math.round(adjPerc * 100)}%`}
              description="Are Adjective"></StatCard>
          </Col>
        </Row>
        <Row className="justify-content-md-center">
          <InputGroup>
            <FormControl
              className="main-text-area"
              as="textarea"
              aria-label="With textarea"
              rows="15"
              placeholder="Enter your text document here ..."
              value={textArea}
              onChange={(e) => {
                setTextArea(e.target.value);
                let wordTok = new natural.WordTokenizer();
                let tokenizedWords = wordTok.tokenize(e.target.value);
                setWordCount(tokenizedWords.length);

                let sentTok = new natural.SentenceTokenizer();
                setSentenceCount(sentTok.tokenize(e.target.value).length);

                let lexicon = new natural.Lexicon("EN", "N", "NNP");
                let ruleSet = new natural.RuleSet("EN");
                let tagger = new natural.BrillPOSTagger(lexicon, ruleSet);
                let taggedWordsSep = tagger.tag(tokenizedWords).taggedWords;

                let nouns = taggedWordsSep.filter(
                  (w) =>
                    w.tag === "NN" ||
                    w.tag === "NNS" ||
                    w.tag === "NNP" ||
                    w.tag === "NNPS"
                );
                let adjs = taggedWordsSep.filter(
                  (w) => w.tag === "JJ" || w.tag === "JJS" || w.tag === "JJR"
                );

                setNounsPerc(nouns.length / taggedWordsSep.length || 0);
                setAdjPerc(adjs.length / taggedWordsSep.length || 0);
              }}
            />
          </InputGroup>
        </Row>
        <Row className="justify-content-md-center">
          <ButtonToolbar aria-label="Toolbar with button groups">
            <ButtonGroup className="mr-2" aria-label="First group">
              <Button className="analyze-button" onClick={() => runAnalysis()}>
                Analyze
              </Button>
            </ButtonGroup>
            <ButtonGroup className="mr-2" aria-label="Second group">
              <Button
                className="analyze-button"
                onClick={() => {
                  setRecording(true);
                  SpeechRecognition.startListening();
                }}>
                <img
                  src="https://img.pngio.com/record-button-png-6-png-image-record-png-2400_2093.png"
                  alt=""
                />
              </Button>
              <Button
                className="analyze-button"
                onClick={() => {
                  resetTranscript();
                  setTextArea("");
                }}>
                Clear
              </Button>
            </ButtonGroup>
          </ButtonToolbar>
        </Row>
        {analysisState !== "DONE" && (
          <Row className="justify-content-md-center TwoTxtSent__row">
            <Col md={1}>
              <Spinner animation="border" role="status"></Spinner>
            </Col>
            <Col md={"auto"}>
              <h3>{analysisState}</h3>
            </Col>
          </Row>
        )}
        {textArea !== "" && (
          <Row>
            <SingleTextAnalysis analysis={textAnalysis}></SingleTextAnalysis>
          </Row>
        )}
      </Container>

      <Modal
        show={recording}
        onHide={() => {
          setRecording(false);
          SpeechRecognition.startListening().then(() => {
            setTextArea(textArea + transcript);
          });
        }}>
        <Modal.Header closeButton></Modal.Header>
        <Modal.Body>Recording ...</Modal.Body>
        <Modal.Footer>
          <Button
            variant="primary"
            onClick={() => {
              setRecording(false);
              SpeechRecognition.startListening().then(() => {
                setTextArea(textArea + transcript);
              });
            }}>
            Finish Recording
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  );
}

export default SingleText;
