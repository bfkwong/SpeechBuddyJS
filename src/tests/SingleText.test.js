import React from "react";
import { render } from "@testing-library/react";
import SingleText from "../components/SingleText/SingleText";
import renderer from "react-test-renderer";
import { unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

describe("SingleText Component", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    if (container.hasChildNodes()) {
      unmountComponentAtNode(container);
    }
    container.remove();
    container = null;
  });

  test("SingleText renders without failure", () => {
    render(<SingleText />, container);
  });

  test("SingleText snapshot test", () => {
    const appTree = renderer.create(<SingleText />).toJSON();
    expect(appTree).toMatchSnapshot();
  });

  test("SingleText Text Field Changes Test", () => {
    let SingleTextComponent = <SingleText />;
    act(() => {
      render(SingleTextComponent, container);
    });

    const SingleTextTA = document.querySelector("[data-testid=SingleTextTA]");
    const SingleTextCLR = document.querySelector("[data-testid=SingleTextCLR]");

    expect(SingleTextTA.placeholder).toBe("Enter your text document here ...");
    expect(SingleTextTA.value).toBe("");

    SingleTextTA.value = "Testing";
    expect(SingleTextTA.value).toBe("Testing");

    SingleTextCLR.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    expect(SingleTextTA.value).toBe("");
  });
});
