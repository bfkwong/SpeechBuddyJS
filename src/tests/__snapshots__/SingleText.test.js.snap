// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleText Component SingleText snapshot test 1`] = `
<div
  className="TopLevelDisplay"
>
  <h1>
    Single Text Analysis
  </h1>
  <div
    className="container"
  >
    <div
      className="justify-content-md-center row"
    >
      <div
        className="col-md-3 col-6"
      >
        <div
          className="card-padding card"
        >
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
              data-testid="SC_value"
            >
              0
            </div>
            <div
              className="mb-2 text-muted card-subtitle h6"
              data-testid="SC_description"
            >
              Words
            </div>
          </div>
        </div>
      </div>
      <div
        className="col-md-3 col-6"
      >
        <div
          className="card-padding card"
        >
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
              data-testid="SC_value"
            >
              1
            </div>
            <div
              className="mb-2 text-muted card-subtitle h6"
              data-testid="SC_description"
            >
              Sentences
            </div>
          </div>
        </div>
      </div>
      <div
        className="col-md-3 col-6"
      >
        <div
          className="card-padding card"
        >
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
              data-testid="SC_value"
            >
              0%
            </div>
            <div
              className="mb-2 text-muted card-subtitle h6"
              data-testid="SC_description"
            >
              Are Nouns
            </div>
          </div>
        </div>
      </div>
      <div
        className="col-md-3 col-6"
      >
        <div
          className="card-padding card"
        >
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
              data-testid="SC_value"
            >
              0%
            </div>
            <div
              className="mb-2 text-muted card-subtitle h6"
              data-testid="SC_description"
            >
              Are Adjective
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="justify-content-md-center row"
    >
      <div
        className="input-group"
      >
        <textarea
          aria-label="With textarea"
          className="main-text-area form-control"
          data-testid="SingleTextTA"
          onChange={[Function]}
          placeholder="Enter your text document here ..."
          rows="15"
          value=""
        />
      </div>
    </div>
    <div
      className="justify-content-md-center row"
    >
      <div
        aria-label="Toolbar with button groups"
        className="btn-toolbar"
        role="toolbar"
      >
        <div
          aria-label="First group"
          className="mr-2 btn-group"
          role="group"
        >
          <button
            className="analyze-button btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Analyze
          </button>
        </div>
        <div
          aria-label="Second group"
          className="mr-2 btn-group"
          role="group"
        >
          <button
            className="analyze-button btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <img
              alt=""
              src="recordbutton.png"
            />
          </button>
          <button
            className="analyze-button btn btn-primary"
            data-testid="SingleTextCLR"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Clear
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
